src/SDK/Trace/Span.php

```diff
@@ -244,19 +248,25 @@ public function setStatus(string $code, ?string $description = null): self
         return $this;
     }
 
-    /** @inheritDoc */
+    /**
+     * @inheritDoc
+     * @psalm-suppress NoInterfaceProperties
+     **/
     public function end(?int $endEpochNanos = null): void
     {
         if ($this->hasEnded) {
             return;
         }
 
         $this->endEpochNanos = $endEpochNanos ?? Clock::getDefault()->now();
-        $this->hasEnded = true;
-
-        $this->checkForDroppedElements();
+        $span = clone $this;
+        $this->hasEnded = true; // prevent further modifications to the span by async code
+        $this->spanProcessor->onEnding($span);
+        $span->hasEnded = true;
+        $span->end();
 
         $this->spanProcessor->onEnd($this);
+        $this->checkForDroppedElements();
```

Comment:
Must use `$span` instead of `$this` to make sure that changes from `::onEnding()` are visible; moving the second callback in `test_span_processor_onending_can_mutate_span` from `::onEnding()` to `::onEnd()` should currently result in a test failure.